src/core/controllers/controller.js

```diff
@@ -0,0 +1,469 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import ViewState from '../lib/view-state';
+import WebMercatorViewport from '../viewports/web-mercator-viewport';
+
+// EVENT HANDLING PARAMETERS
+const ZOOM_ACCEL = 0.01;
+
+const PITCH_MOUSE_THRESHOLD = 5;
+const PITCH_ACCEL = 1.2;
+
+const EVENT_TYPES = {
+  WHEEL: ['wheel'],
+  PAN: ['panstart', 'panmove', 'panend'],
+  PINCH: ['pinchstart', 'pinchmove', 'pinchend'],
+  DOUBLE_TAP: ['doubletap'],
+  KEYBOARD: ['keydown', 'keyup']
+};
+
+const noop = viewState => viewState;
+
+const DEFAULT_EVENT_REDUCERS = {
+  panStart: noop,
+  pan: noop,
+  panEnd: noop,
+  rotateStart: noop,
+  rotate: noop,
+  rotateEnd: noop,
+  zoomStart: noop,
+  zoom: noop,
+  zoomEnd: noop,
+  zoomIn: noop,
+  zoomOut: noop,
+  moveLeft: noop,
+  moveRight: noop,
+  moveForward: noop,
+  moveBackward: noop
+};
+
+const KEY_TO_REDUCER_BINDINGS = {
+  w: 'moveForward',
+  W: 'moveForward',
+  ArrowUp: 'moveForward',
```

Comment:
In first person experience, arrow keys usually move the lookAt and ADWS keys move the camera. Should at least differentiate them in the base class.